pcdsdevices/positioner.py

```diff
@@ -0,0 +1,187 @@
+    """
+    def __init__(
+            self, *, name, move, get_pos, set_pos=None, stop=None, done=None,
+            check_value=None, info=None, egu='', limits=None, update_rate=1,
+            timeout=60, notepad_pv=None, parent=None, kind=None,
+            **kwargs
+            ):
+        self._move = move
+        self._get_pos = get_pos
+        self._set_pos = set_pos
+        self._stop = stop
+        self._done = done
+        self._check = check_value
+        self._info = info
+        self._last_update = 0
+        self._goal = None
+        self.update_rate = 1
+        notepad_name = name + '_notepad'
+        if notepad_pv is None:
+            self.notepad_signal = Signal(name=notepad_name)
+        else:
+            self.notepad_signal = EpicsSignal(notepad_pv, name=notepad_name)
+        if parent is None and kind is None:
+            kind = 'hinted'
+        super().__init__(name=name, egu=egu, limits=limits, source='func',
+                         timeout=timeout, parent=parent, kind=kind, **kwargs)
+
+    def _setup_move(self, position, status):
+        self._run_subs(sub_type=self.SUB_START, timestamp=time.time())
+        self._goal = position
+        self._started_moving = True
+        self._moving = True
+        self._finished = False
+        self._move(position)
+        self._new_update(status)
+
+    def _new_update(self, status):
+        schedule_task(
+            self._update_task, args=(status,), delay=self.update_rate
+            )
+
+    def _update_task(self, status):
+        self._update_position()
+        if self._check_finished():
+            try:
+                status.set_finished()
+            except InvalidState:
+                pass
+        if status.done:
+            self._started_moving = False
+            self._moving = False
+        else:
+            self._new_update(status)
+
+    @property
+    def position(self):
+        self._update_position()
+        return self._position
+
+    def _update_position(self):
+        if time.time() - self._last_update > self.update_rate:
+            self._last_update = time.time()
+            pos = self._get_pos()
+            self._set_position(pos)
+            self.notepad_signal.put(pos)
+
+    def _check_finished(self):
+        if self._done is None:
+            finished = np.isclose(self._goal, self.position)
+        else:
+            finished = self._done()
+        if finished:
+            self._done_moving()
+        return finished
+
+    def set_position(self, position):
+        if self._set_pos is None:
+            raise NotImplementedError(
+                f'FuncPositioner {self.name} was not '
+                'given a set_pos argument.'
+                )
+        else:
+            self._set_pos(position)
+
+    def stop(self, *args, **kwargs):
+        if self._stop is None:
+            # Often called by bluesky, don't throw an exception
+            self.log.warning(
```

Comment:
This will probably be annoying to users, but rightfully so. Hopefully they'll read it and pass in a stop function.

Reply:
It probably only shows up when a bluesky scan pauses or has other issues

Reply:
Decided to keep it fairly simple: https://github.com/pcdshub/pcdsdevices/pull/786/commits/b5aee231e5040d247aaa633059b8478e9763d54b